uaclient/tests/test_config.py

```diff
@@ -1,202 +1,195 @@
+        cfg = UAConfig({'data_dir': tmpdir.strpath})
         cfg.write_cache('accounts', {'non-accounts': 'somethingelse'})
 
         assert [] == cfg.accounts
         expected_warning = (
-            "WARNING: Missing 'accounts' key in cache %s" %
-            self.tmp_path('accounts.json', tmp_dir))
-        assert expected_warning in self.logs
+            "WARNING  Missing 'accounts' key in cache %s" %
+            tmpdir.join('accounts.json'))
+        assert expected_warning in caplog_text()
 
-    def test_accounts_logs_warning_when_non_list_accounts_cache_content(self):
+    def test_accounts_logs_warning_when_non_list_accounts_cache_content(
+            self, caplog_text, tmpdir):
         """Config.accounts warns on non-list accounts key."""
-        tmp_dir = self.tmp_dir()
-        cfg = UAConfig({'data_dir': tmp_dir})
+        cfg = UAConfig({'data_dir': tmpdir.strpath})
         cfg.write_cache('accounts', {'accounts': 'non-list-value'})
 
         assert [] == cfg.accounts
         expected_warning = (
-            "WARNING: Unexpected 'accounts' type <class 'str'> in cache %s" % (
-                self.tmp_path('accounts.json', tmp_dir)))
-        assert expected_warning in self.logs
+            "WARNING  Unexpected 'accounts' type <class 'str'> in cache %s" % (
+                tmpdir.join('accounts.json')))
+        assert expected_warning in caplog_text()
 
 
-class TestDataPath(TestCase):
+class TestDataPath:
 
     def test_data_path_returns_data_dir_path_without_key(self):
         """The data_path method returns the data_dir when key is absent."""
         cfg = UAConfig({'data_dir': '/my/dir'})
         assert '/my/dir' == cfg.data_path()
 
-    @params(*KNOWN_DATA_PATHS)
+    @pytest.mark.parametrize('key,path_basename', KNOWN_DATA_PATHS)
     def test_data_path_returns_file_path_with_defined_data_paths(
             self, key, path_basename):
         """When key is defined in Config.data_paths return data_path value."""
         cfg = UAConfig({'data_dir': '/my/dir'})
         assert '/my/dir/%s' % path_basename == cfg.data_path(key=key)
 
-    @params(('notHere', 'notHere'), ('anything', 'anything'))
+    @pytest.mark.parametrize('key,path_basename', (
+        ('notHere', 'notHere'), ('anything', 'anything')))
     def test_data_path_returns_file_path_with_undefined_data_paths(
             self, key, path_basename):
         """When key is not in Config.data_paths the key is used to data_dir"""
         cfg = UAConfig({'data_dir': '/my/dir'})
         assert '/my/dir/%s' % key == cfg.data_path(key=key)
 
 
-class TestWriteCache(TestCase):
+class TestWriteCache:
 
-    @params(('unknownkey', 'content1'), ('another-one', 'content2'))
+    @pytest.mark.parametrize('key,content', (
+        ('unknownkey', 'content1'), ('another-one', 'content2')))
     def test_write_cache_write_key_name_in_data_dir_when_data_path_absent(
-            self, key, content):
+            self, tmpdir, key, content):
```

Comment:
This test is a useful example of the `tmpdir` fixture; it doesn't inject a string containing a path, it injects a `py.path` object. These have a rich set of methods that allow for more expressive tests.

In this test, for example:
* we use `tmpdir.strpath` to get a string containing the path to the temporary directory
* we use `tmpdir.join` to produce a new `py.path` object for a sub-path of our temporary directory
* we use `.check()` on that constructed object to check whether or not it exists

Reply:
Excellent inline comments here for context. Thanks for the good examples and doc references